src/Vesting.ts

```diff
@@ -0,0 +1,208 @@
+import { StellarTxnContext } from "../lib/StellarTxnContext";
+
+import {ADA} from "../lib/StellarTestHelper";
+
+export type VestingParams = {
+    sponsor: WalletEmulator;
+    payee: Address;
+    deadline: bigint;
+};
+
+export type VestingDatumArgs = {
+    sponsor: PubKeyHash;
+    payee: PubKeyHash;
+    time: number | bigint;
+};
+
+export class Vesting extends StellarContract<VestingParams> {
+    contractSource() {
+        return contract;
+    }
+    @datum
+    mkDatum({
+        sponsor,
+        payee,
+	time
+    }: VestingDatumArgs): InlineDatum {
+        //!!! todo: make it possible to type these datum helpers more strongly
+        const t = new this.configuredContract.types.Datum(
+            sponsor.bytes,
+	    payee.bytes,
+            time
+        );
+        return Datum.inline(t._toUplcData());
+    }
+    @txn
+    async mkTxnDepositValueForVesting(
+        { sponsor, payee, deadline }: VestingParams,
+        tcx: StellarTxnContext = new StellarTxnContext()
+    ): Promise<StellarTxnContext | never> {
+	    // so far value is hardcoded: 
+	    	const margin = 5n * ADA;
+		const inUtxo = (await sponsor.utxos)[0];
+		const inUtxoFee = (await sponsor.utxos)[1];
+		const lockedVal = inUtxo.value; // TODO: parametrize
+
+		const validatorAddress = Address.fromValidatorHash(this.compiledContract.validatorHash)
+
+		const inlineDatum = this.mkDatum({
+			sponsor: sponsor.address.pubKeyHash,
+			payee: payee.pubKeyHash,
+			time: deadline
+		});
+
+		tcx.addInput(inUtxo)
+		   .addInput(inUtxoFee)
+		   .addOutput(new TxOutput(validatorAddress, lockedVal, inlineDatum))
+                   .addOutput(
+                    new TxOutput(
+                        sponsor.address,
+                        new Value(inUtxoFee.value.lovelace - margin)
+                    )
+                );
+	return tcx
+    }
+    @txn
+    async mkTxnCancelVesting(
+	sponsor: WalletEmulator,
+	valUtxo: UTxO,
+	validFrom: Date,
+	validTill: Date,
+        tcx: StellarTxnContext = new StellarTxnContext()
+    ): Promise<StellarTxnContext | never> {
+	    // How does it work?
+	    // It creates a Redeemer and serializes it:
+	   const r = new this.configuredContract.types.Redeemer.Cancel();
+	   const valRedeemer = r._toUplcData();
+
+	   // finds enough utxos:
+	   const collateralUtxo = (await sponsor.utxos)[0];
+	   const feeUtxo = (await sponsor.utxos)[1];
+
+	   // Calculates validity interval:
+           // const validFrom = h.currentSlot() - 1n;
+	   // const validTo = h.currentSlot() + 500n;
+
+	   //creates the transaction and adds its components:
+	   tcx.addInput(feeUtxo)
+	   	.addInput(valUtxo, valRedeemer)
+           	.addOutput(new TxOutput(sponsor.address, valUtxo.value))
+           	
+           	.attachScript(this.compiledContract)
+           	.addCollateral(collateralUtxo);
+	tcx.tx.addSigner(sponsor.address.pubKeyHash);
+	// need to pass both, see junk/dev1
+	tcx.tx.validFrom(validFrom);
+	tcx.tx.validTo(validTill);
```

Comment:
https://www.hyperion-bt.org/helios-book/api/reference/classes/Tx.html?highlight=valid#validfrom

Set the start of the valid time range by specifying either a Date or a slot.

Since 13.29 is set automatically. What if I remove it?

Reply:
Error: tx invalid (not finalized or slot out of range)